File: backend/conflict_precheck.py
"""Heuristic conflict pre-check.

Python port of the legacy Node prototype's ``src/conflict-precheck.mjs``. Flags
likely-conflicting requirement pairs by regex/keyword heuristics before an LLM
ever sees them, and reconciles an LLM review's conflict list against those
candidates so conflicts the model missed are not silently dropped.
"""

import re
from typing import Any, Dict, List, Optional, TypedDict


class Requirement(TypedDict):
    id: str
    text: str


class ConflictCandidate(TypedDict):
    id: str
    requirementIds: List[str]
    type: str
    severity: str
    evidence: str
    mitigation: str


NEGATIVE_PATTERNS = [
    re.compile(r"\bshall not\b", re.IGNORECASE),
    re.compile(r"\bmust not\b", re.IGNORECASE),
    re.compile(r"\bwill not\b", re.IGNORECASE),
    re.compile(r"\bnot\b", re.IGNORECASE),
    re.compile(r"\bnever\b", re.IGNORECASE),
    re.compile(r"\bprohibit", re.IGNORECASE),
    re.compile(r"\bdisable", re.IGNORECASE),
]

POSITIVE_PATTERNS = [
    re.compile(r"\bshall\b", re.IGNORECASE),
    re.compile(r"\bmust\b", re.IGNORECASE),
    re.compile(r"\bwill\b", re.IGNORECASE),
    re.compile(r"\btransmit\b", re.IGNORECASE),
    re.compile(r"\bprovide\b", re.IGNORECASE),
    re.compile(r"\benable\b", re.IGNORECASE),
    re.compile(r"\ballow\b", re.IGNORECASE),
]

STOPWORDS = {
    "the",
    "and",
    "shall",
    "must",
    "will",
    "when",
    "while",
    "with",
    "from",
    "between",
    "greater",
    "than",
    "less",
    "system",
    "data",
}

_NUMBER_PATTERN = re.compile(r"(.{0,25}?)(\d+(?:\.\d+)?)\s*(km/h|ms|s|v|a|%|lux|hz)?", re.IGNORECASE)
_MAX_CONTEXT = re.compile(r"within|less than|below|under|max|maximum|no more than", re.IGNORECASE)
_MIN_CONTEXT = re.compile(r"greater than|above|over|min|minimum|at least", re.IGNORECASE)


def find_conflict_candidates(requirements: List[Requirement]) -> List[ConflictCandidate]:
    candidates: List[ConflictCandidate] = []
    for left_index in range(len(requirements)):
        for right_index in range(left_index + 1, len(requirements)):
            left = requirements[left_index]
            right = requirements[right_index]
            similarity = _jaccard(_keywords(left["text"]), _keywords(right["text"]))
            left_negative = _matches_any(left["text"], NEGATIVE_PATTERNS)
            right_negative = _matches_any(right["text"], NEGATIVE_PATTERNS)
            left_positive = _matches_any(left["text"], POSITIVE_PATTERNS)
            right_positive = _matches_any(right["text"], POSITIVE_PATTERNS)

            if similarity >= 0.3 and left_negative != right_negative and left_positive and right_positive:
                candidates.append(
                    {
                        "id": f"PC-{len(candidates) + 1:03d}",
                        "requirementIds": [left["id"], right["id"]],
                        "type": "contradiction",
                        "severity": "high",
                        "evidence": (
                            f"{left['id']} requires the behavior while {right['id']} prohibits the "
                            "same or overlapping behavior."
                        ),
                        "mitigation": (
                            "Clarify the applicable mode, speed range, operating condition, or "
                            "product variant; otherwise choose one behavior and update the "
                            "conflicting requirement."
                        ),
                    }
                )
                continue

            threshold = _threshold_conflict(left, right, similarity)
            if threshold:
                candidates.append(
                    {
                        "id": f"PC-{len(candidates) + 1:03d}",
                        "requirementIds": [left["id"], right["id"]],
                        "type": "threshold mismatch",
                        "severity": "medium",
                        "evidence": threshold,
                        "mitigation": (
                            "Align the threshold values or split the requirements by interface, "
                            "mode, or variant."
                        ),
                    }
                )
    return candidates


def merge_candidate_conflicts(review: Dict[str, Any], candidates: List[ConflictCandidate]) -> Dict[str, Any]:
    existing_keys = {_conflict_key(conflict) for conflict in review["conflicts"]}
    missing = [candidate for candidate in candidates if _conflict_key(candidate) not in existing_keys]
    conflict_ids = {
        requirement_id
        for candidate in candidates
        for requirement_id in candidate["requirementIds"]
    }

    requirements = []
    for requirement in review["requirements"]:
        if requirement["id"] not in conflict_ids:
            requirements.append(requirement)
            continue

        score = 75 if requirement["score"] == 0 else min(requirement["score"], 82)
        flags = list(dict.fromkeys([*requirement.get("flags", []), "set-level conflict"]))
        requirements.append(
            {
                **requirement,
                "score": score,
                "flags": flags,
                "improvement": requirement.get("improvement")
                or "Resolve the set-level conflict by clarifying mode, condition, threshold, or selected behavior.",
            }
        )

    average_score = round(sum(r["score"] for r in requirements) / len(requirements)) if requirements else 0
    below_threshold_count = sum(1 for r in requirements if r["score"] < 85)

    return {
        **review,
        "summary": {
            **review["summary"],
            "averageScore": average_score,
            "belowThresholdCount": below_threshold_count,
            "conflictCount": len(review["conflicts"]) + len(missing),
        },
        "requirements": requirements,
        "conflicts": [*review["conflicts"], *missing],
    }


def _conflict_key(conflict: Dict[str, Any]) -> str:
    return "|".join(sorted(conflict["requirementIds"]))


def _threshold_conflict(left: Requirement, right: Requirement, similarity: float) -> str:
    if similarity < 0.35:
        return ""

    left_numbers = _numbers_with_units(left["text"])
    right_numbers = _numbers_with_units(right["text"])
    for left_number in left_numbers:
        for right_number in right_numbers:
            if left_number["unit"] != right_number["unit"] or left_number["operator"] != right_number["operator"]:
                continue
            delta = abs(left_number["value"] - right_number["value"])
            if delta > max(left_number["value"], right_number["value"]) * 0.25:
                return (
                    f"{left['id']} and {right['id']} use different {left_number['operator']} values: "
                    f"{left_number['value']}{left_number['unit']} vs {right_number['value']}{right_number['unit']}."
                )
    return ""


def _numbers_with_units(text: str) -> List[Dict[str, Any]]:
    results = []
    for match in _NUMBER_PATTERN.finditer(text):
        raw_unit = (match.group(3) or "").lower()
        results.append(
            {
                "value": _canonical_value(float(match.group(2)), raw_unit),
                "unit": _canonical_unit(raw_unit),
                "operator": _operator_from_context(match.group(1)),
            }
        )
    return results


def _canonical_value(value: float, unit: str) -> float:
    return value / 1000 if unit == "ms" else value


def _canonical_unit(unit: str) -> str:
    return "s" if unit == "ms" else unit


def _operator_from_context(context: str) -> str:
    if _MAX_CONTEXT.search(context):
        return "maximum"
    if _MIN_CONTEXT.search(context):
        return "minimum"
    return "value"


def _matches_any(text: str, patterns: List[re.Pattern]) -> bool:
    return any(pattern.search(text) for pattern in patterns)


def _keywords(text: Optional[str]) -> set:
    words = re.findall(r"[a-z0-9]+", str(text or "").lower())
    return {word for word in words if len(word) > 2 and word not in STOPWORDS}


def _jaccard(left: set, right: set) -> float:
    if not left or not right:
        return 0.0
    intersection = len(left & right)
    union = len(left | right)
    return intersection / union

File: legacy/node-prototype/src/conflict-precheck.mjs
const NEGATIVE_PATTERNS = [
  /\bshall not\b/i,
  /\bmust not\b/i,
  /\bwill not\b/i,
  /\bnot\b/i,
  /\bnever\b/i,
  /\bprohibit/i,
  /\bdisable/i
];

const POSITIVE_PATTERNS = [
  /\bshall\b/i,
  /\bmust\b/i,
  /\bwill\b/i,
  /\btransmit\b/i,
  /\bprovide\b/i,
  /\benable\b/i,
  /\ballow\b/i
];

const STOPWORDS = new Set([
  "the",
  "and",
  "shall",
  "must",
  "will",
  "when",
  "while",
  "with",
  "from",
  "between",
  "greater",
  "than",
  "less",
  "system",
  "data"
]);

export function findConflictCandidates(requirements) {
  const candidates = [];
  for (let leftIndex = 0; leftIndex < requirements.length; leftIndex += 1) {
    for (let rightIndex = leftIndex + 1; rightIndex < requirements.length; rightIndex += 1) {
      const left = requirements[leftIndex];
      const right = requirements[rightIndex];
      const similarity = jaccard(keywords(left.text), keywords(right.text));
      const leftNegative = matchesAny(left.text, NEGATIVE_PATTERNS);
      const rightNegative = matchesAny(right.text, NEGATIVE_PATTERNS);
      const leftPositive = matchesAny(left.text, POSITIVE_PATTERNS);
      const rightPositive = matchesAny(right.text, POSITIVE_PATTERNS);

      if (similarity >= 0.3 && leftNegative !== rightNegative && leftPositive && rightPositive) {
        candidates.push({
          id: `PC-${String(candidates.length + 1).padStart(3, "0")}`,
          requirementIds: [left.id, right.id],
          type: "contradiction",
          severity: "high",
          evidence: `${left.id} requires the behavior while ${right.id} prohibits the same or overlapping behavior.`,
          mitigation: "Clarify the applicable mode, speed range, operating condition, or product variant; otherwise choose one behavior and update the conflicting requirement."
        });
        continue;
      }

      const threshold = thresholdConflict(left, right, similarity);
      if (threshold) {
        candidates.push({
          id: `PC-${String(candidates.length + 1).padStart(3, "0")}`,
          requirementIds: [left.id, right.id],
          type: "threshold mismatch",
          severity: "medium",
          evidence: threshold,
          mitigation: "Align the threshold values or split the requirements by interface, mode, or variant."
        });
      }
    }
  }
  return candidates;
}

export function mergeCandidateConflicts(review, candidates) {
  const existingKeys = new Set(review.conflicts.map(conflictKey));
  const missing = candidates.filter(candidate => !existingKeys.has(conflictKey(candidate)));
  const conflictIds = new Set(candidates.flatMap(candidate => candidate.requirementIds));
  const requirements = review.requirements.map(requirement => {
    if (!conflictIds.has(requirement.id)) {
      return requirement;
    }

    const score = requirement.score === 0 ? 75 : Math.min(requirement.score, 82);
    const flags = new Set([...(requirement.flags || []), "set-level conflict"]);
    return {
      ...requirement,
      score,
      flags: [...flags],
      improvement: requirement.improvement || "Resolve the set-level conflict by clarifying mode, condition, threshold, or selected behavior."
    };
  });
  const averageScore = requirements.length
    ? Math.round(requirements.reduce((sum, requirement) => sum + requirement.score, 0) / requirements.length)
    : 0;
  const belowThresholdCount = requirements.filter(requirement => requirement.score < 85).length;

  return {
    ...review,
    summary: {
      ...review.summary,
      averageScore,
      belowThresholdCount,
      conflictCount: review.conflicts.length + missing.length
    },
    requirements,
    conflicts: [...review.conflicts, ...missing]
  };
}

function conflictKey(conflict) {
  return [...conflict.requirementIds].sort().join("|");
}

function thresholdConflict(left, right, similarity) {
  if (similarity < 0.35) {
    return "";
  }

  const leftNumbers = numbersWithUnits(left.text);
  const rightNumbers = numbersWithUnits(right.text);
  for (const leftNumber of leftNumbers) {
    for (const rightNumber of rightNumbers) {
      if (leftNumber.unit !== rightNumber.unit || leftNumber.operator !== rightNumber.operator) {
        continue;
      }
      const delta = Math.abs(leftNumber.value - rightNumber.value);
      if (delta > Math.max(leftNumber.value, rightNumber.value) * 0.25) {
        return `${left.id} and ${right.id} use different ${leftNumber.operator} values: ${leftNumber.value}${leftNumber.unit} vs ${rightNumber.value}${rightNumber.unit}.`;
      }
    }
  }
  return "";
}

function numbersWithUnits(text) {
  const matches = [...text.matchAll(/(.{0,25}?)(\d+(?:\.\d+)?)\s*(km\/h|ms|s|v|a|%|lux|hz)?/gi)];
  return matches.map(match => ({
    value: canonicalValue(Number(match[2]), String(match[3] || "").toLowerCase()),
    unit: canonicalUnit(String(match[3] || "").toLowerCase()),
    operator: operatorFromContext(match[1])
  }));
}

function canonicalValue(value, unit) {
  return unit === "ms" ? value / 1000 : value;
}

function canonicalUnit(unit) {
  return unit === "ms" ? "s" : unit;
}

function operatorFromContext(context) {
  if (/within|less than|below|under|max|maximum|no more than/i.test(context)) {
    return "maximum";
  }
  if (/greater than|above|over|min|minimum|at least/i.test(context)) {
    return "minimum";
  }
  return "value";
}

function matchesAny(text, patterns) {
  return patterns.some(pattern => pattern.test(text));
}

function keywords(text) {
  return new Set(
    String(text || "")
      .toLowerCase()
      .match(/[a-z0-9]+/g)
      ?.filter(word => word.length > 2 && !STOPWORDS.has(word)) || []
  );
}

function jaccard(left, right) {
  if (!left.size || !right.size) {
    return 0;
  }
  const intersection = [...left].filter(word => right.has(word)).length;
  const union = new Set([...left, ...right]).size;
  return intersection / union;
}
